unix: Fix support for Bluetooth via libusb, add HCI dump feature

yasnim24 opened the pull request:
ports/unix did not build with
"make MICROPY_PY_BLUETOOTH=1 MICROPY_BLUETOOTH_BTSTACK=1".

This PR fixes it. If libusb-1.0 exists on the machine "MICROPY_BLUETOOTH_BTSTACK_USB=1" is
automatically set and Bluetooth over USB should work.

Now also a log file "/tmp/hci_dump.pklg" could be activated by setting
"MICROPY_BLUETOOTH_BTSTACK_HCI_DUMP=1".

yasnim24 commented:
Several minutes ago I saw a commit that could be used to replace the init_mp_state_thread() function in "ports/unix/mpbtstackport_usb.c":
commit bc424dd: "py/modthread: Move thread state initialisation to shared function."

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

yasnim24 committed:
ports/unix: Change 'STATIC' to 'static'.

See 'global: Remove the STATIC macro.':
https://github.com/micropython/micropython/pull/13763

Signed-off-by: Yasmin Bosch <development@yasmin-bosch.de>

yasnim24 committed:
ports/unix: Use 'mp_thread_init_state()' instead of own function.

The function 'mp_thread_init_state()' has been added to 'py/runtime.h'
with commit #bc424dd in micropython/micropython:
'py/modthread: Move thread state initialisation to shared function.'

So use that function instead of defining own function in
'ports/unix/mpbtstackport_usb.c'.

Signed-off-by: Yasmin Bosch <development@yasmin-bosch.de>

yasnim24 committed:
ports/unix/Makefile: Move up "HAVE_LIBUSB" code to be effective.

There is a code that evaluates "HAVE_LIBUSB" and then sets
"MICROPY_BLUETOOTH_BTSTACK_USB" and "MICROPY_BLUETOOTH_BTSTACK_H4"
accordingly.
That code has been moved before "include $(TOP)/extmod/extmod.mk"
to be effective.

Details:
"include $(TOP)/extmod/extmod.mk" will
"include $(TOP)/extmod/btstack/btstack.mk"
and that in turn sets
"MICROPY_BLUETOOTH_BTSTACK_USB ?= 0"
"MICROPY_BLUETOOTH_BTSTACK_H4 ?= 1"

Signed-off-by: Yasmin Bosch <development@yasmin-bosch.de>

yasnim24 committed:
ports/unix: Fix "‘signal_sco_can_send_now’ defined but not used".

Added a "ports/unix/btstack_config.h" for a workaround:
If "MICROPY_BLUETOOTH_BTSTACK_USB" is set and "ENABLE_SCO_OVER_HCI"
is not defined there is a compiler error:
    ../../lib/btstack/platform/libusb/hci_transport_h2_libusb.c:1354:13:
    error: ‘signal_sco_can_send_now’ defined but not used
Because the problem is in an external code a workaround is done by
defining MICROPY_BLUETOOTH_BTSTACK_USB in the added "btstack_config.h".

The "ports/unix/Makefile" has been changed so that the "btstack_config.h"
will be included.

Signed-off-by: Yasmin Bosch <development@yasmin-bosch.de>

yasnim24 committed:
ports/unix: Fix so that "MICROPY_BLUETOOTH_BTSTACK_USB=1" is working.

ports/unix/mpbtstackport_usb.c:
"lib/btstack/platform/posix/btstack_run_loop_posix.c"
is now used instead of
"lib/btstack/platform/embedded/btstack_run_loop_embedded.c".
Additional code has been added to make it work.
Now also a log file "/tmp/hci_dump.pklg" could be activated by setting
"MICROPY_BLUETOOTH_BTSTACK_HCI_DUMP=1".

The "Makefile" and "mpbtstackport_common.c" has been changed accordingly.

Signed-off-by: Yasmin Bosch <development@yasmin-bosch.de>

yasnim24 commented:
The first three commits #29f6e66, #7acadbd, #5393f63 are identical with the original commits except that the <author> has been fixed.
The other two commits #b4c69d8 and #0e75c12 are new.

projectgus commented on ports/unix/Makefile:
```diff
@@ -144,21 +169,27 @@ ifeq ($(MICROPY_SSL_AXTLS),1)
-else
-# Fallback to HCI controller via a H4 UART (e.g. Zephyr on nRF) over a /dev/tty serial port.
-MICROPY_BLUETOOTH_BTSTACK_H4 ?= 1
+ifndef MICROPY_BLUETOOTH_BTSTACK_HCI_DUMP
```
Please add a comment here about what this value is for, similar to the other config items.

Suggest even moving it up with the other user-settable values.

projectgus commented on ports/unix/Makefile:
```diff
@@ -33,6 +33,31 @@ QSTR_GLOBAL_DEPENDENCIES += $(VARIANT_DIR)/mpconfigvariant.h
+# 
+# The original file "extmod/btstack/btstack_config_common.h"
+# will be included by it. 
+MICROPY_BLUETOOTH_BTSTACK_CONFIG_FILE ?= '"ports/unix/btstack_config.h"'
```
Suggest moving this up before the `endif` for bluetooth

projectgus commented on ports/unix/Makefile:
```diff
@@ -33,6 +33,31 @@ QSTR_GLOBAL_DEPENDENCIES += $(VARIANT_DIR)/mpconfigvariant.h
+endif
+
+endif
+endif
```
```suggestion
endif  # MICROPY_PY_BLUETOOTH
```

(I know the original version doesn't have this, but it makes it all a lot easier to follow.)

projectgus reviewed (commented):
Thanks for submitting this PR, @yasnim24. 

I was a little confused by the title - the unix port already has support for Bluetooth via libusb, but it's broken on master and this PR fixes it, yes? The only new functionality is the HCI dump option.

I think the changes look reasonable to me. We should probably add a bluetooth variant (or two) to the unix port and build them in CI, though, to catch any future bitrot of this kind. You don't have to do that part unless you're keen, though.

yasnim24 commented:
Thank you for your comment, @projectgus.

Your suggested changes are OK for me.
But I'm not sure what to do. Should I add new commits for the changes? Or do I have to change the original commits and do a force push?

Can you also further explain what you mean with:
> ... We should probably add a bluetooth variant (or two) to the unix port and build them in CI, though, to catch any future bitrot of this kind. ...

projectgus commented:
> But I'm not sure what to do. Should I add new commits for the changes? Or do I have to change the original commits and do a force push?

Best to rebase, change the commits, and force push to keep clean commit history. Jimmo has pointers published at https://github.com/jimmo/git-and-micropython#README

> Can you also further explain what you mean with:
> 
> > ... We should probably add a bluetooth variant (or two) to the unix port and build them in CI, though, to catch any future bitrot of this kind. ...

We have some "variants" for the unix port with different configs:
https://github.com/micropython/micropython/tree/master/ports/unix/variants

However I don't think any of the variants enable the Bluetooth feature. Therefore, when the code broke ("bitrotted") at some point in the past nobody noticed until now. If we had a variant to build Bluetooth via GitHub Actions (Continuous Integration aka CI) then we would have known that this feature no longer compiled.

Don't worry about that for this PR, though.

yasnim24 commented:
Merged changes as suggested by @projectgus to commits [afe2867](https://github.com/micropython/micropython/pull/14006/commits/afe286737f19b3a5bbcdfc452abfb5224db8137d), [60a5e47](https://github.com/micropython/micropython/pull/14006/commits/60a5e47114769f5c20713b48b9b68829e45b80ed).

projectgus reviewed (approved):
Sorry for the long delay for re-review, @yasnim24 . This looks good to me, and useful!

dpgeorge commented:
@agatti your PR #18972 does fix the SCO gcc warning in a cleaner way.  But it seems there are still some improvements in this PR which yours does not cover.

It would be good to merge both this and your PR, but there are probably going to be clashes.  What do you think?

agatti commented:
I was planning to rework this into at least two commits, one for the build fix and the other one for the HCI dump.  Since I've got no control over this branch I'll probably add the necessary commits into my repo whilst keeping the original authorship information.

If my SCO fix looks nicer, then I'll keep that and retain the rest.  Let me work on this a bit and see how it ends up looking.

agatti commented:
OK, I've pushed the modified commits to https://github.com/agatti/micropython/tree/bt-usb-fix-tmp.

I've split the patch into three commits, each using commit subjects that follow the repo guidelines:

- Build fix: mostly the same as the original, with my SCO compilation fix.
- New event loop for BT-over-USB: `pkg-config` is searched using `command -v` first, as per POSIX specs, made the stack size configurable, replaced `log_info` calls with `DEBUG_printf`
- HCI dump: the packet file name directory is looked up using `$TMPDIR` first, and if that's not the case then it defaults to `/tmp`, plus if `$MICROPY_BLUETOOTH_HCI_DUMP` is set, it will use that as the file path.

I've retained the authorship and reused as much of the original commit messages as possible - timestamps may not have been properly preserved.

If I didn't butcher the original PR too much, let me know so those commits can be cherry-picked as they are or if I have to replace my PR's content with these new commits.

dpgeorge commented:
@agatti thanks for doing that.  Your 3 commits look good, except that it seems you missed the 5th commit from this PR, which removes `init_mp_state_thread()` and instead uses `mp_thread_init_state()`.  Was there a reason to leave that out?

agatti commented:
It was probably too late at night and I missed that :|

I've merged commit 5 of the original PR into commit 2 of the new patch set.  The commit message doesn't really need to be changed, I think.
